Esql/create DATE_DIFF function

nicolasgras opened the pull request:
This pull request resolves https://github.com/elastic/elasticsearch/issues/101942.

A new functionality DATE_DIFF is added to [ESQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql.html). This functionaliy is similar to the ones in [SQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-diff) plugin.

It calculates the diffenrence between this 2 dates using a unit specified and takes 3 input parameters:
- unit (text)
- startTimestamp (dateTime)
- endTimestamp (dateTime)

![Screenshot from 2023-12-08 01-37-41](https://github.com/elastic/elasticsearch/assets/8484741/3b612758-5ad1-412e-921a-a9f86abe80d3)

nicolasgras committed:
Create DateDiff class in ESQL plugin

nicolasgras committed:
Add date_diff function in ESQL plugin

nicolasgras committed:
Update date_diff function in csv-spec

nicolasgras committed:
Update doc with date_diff in ESQL

nicolasgras committed:
Add changelog file for issue 101942

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+        if (datePartField == null) {
+            List<String> similar = Part.findSimilar(unit.utf8ToString());
+            if (similar.isEmpty()) {
+                throw new InvalidArgumentException(
```
A nit on style: you could prepare the message in the two branches and throw just once.

nicolasgras replied:
I agree, I have updated the code.

bpintea reviewed (commented):
Thanks for your interest and contribution, @nicolasgras.
I had a first pass. My main question is about the code duplication. Any reason you didn't extract the common parts (in the ql module) and reused it here?

nicolasgras committed:
Update code after code review

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+            return aliases;
+        }
+
+        private static int safeInt(long diff) {
```
I have updated the code to use DataTypeConverter#safeToInt and removed the one I implemented that was identical.

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+
+        static {
+            NAME_TO_PART = DateTimeField.initializeResolutionMap(values());
+            VALID_VALUES = DateTimeField.initializeValidValues(values());
```
I agree, VALID_VALUES List is not used. I should have cleant this part.
The initialization of NAME_TO_PART is inlined now.

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+
+        private BiFunction<ZonedDateTime, ZonedDateTime, Integer> diffFunction;
+        private Set<String> aliases;
+
```
I agree, they are final now.

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+        description = "Subtract 2 dates and return their difference in multiples of a unit specified in the 1st argument")
+    public DateDiff(
+        Source source,
+        @Param(name = "unit", type = { "text" }, description = "A valid date unit") Expression unit,
```
I agree, I have updated the code.

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+        this.endTimestamp = endTimestamp;
+    }
+
+    @Evaluator
```
I agree, I have updated the code.

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+        ZonedDateTime zdtStart = ZonedDateTime.ofInstant(Instant.ofEpochMilli(startTimestamp), UTC);
+        ZonedDateTime zdtEnd = ZonedDateTime.ofInstant(Instant.ofEpochMilli(endTimestamp), UTC);
+
+        Part datePartField = Part.resolve(unit.utf8ToString());
```
I have updated the code by adding a new evaluator (I get something similar to DateParse now).

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,216 @@
+            type = { "date" },
+            description = "A string representing a start timestamp"
+        ) Expression startTimestamp,
+        @Param(name = "endTimestamp", type = { "date" }, description = "A string representing a end timestamp") Expression endTimestamp
```
This typo has been fixed.

nicolasgras commented:
> Thanks for your interest and contribution, @nicolasgras. I had a first pass. My main question is about the code duplication. Any reason you didn't extract the common parts (in the ql module) and reused it here?

Thank you for you review @bpintea.
As this issue is about **esql**, I thought I should not update any other plugin. But it is better to use it in a common module.
As you suggest, I have moved **DateTimeFiled** interface in **ql** module. Maybe let's keep in mind to also use it in sql module if this PR is accepted and merged.

I have also added unit tests in esql/qa/testFixtures directory to have a better code coverage.

Let me know if it is ok for you.

luigidellaquila reviewed (commented):
Thank you very much @nicolasgras!
Just two comments from my side:
- I see the CI still complains about CLA signature; I checked your commits and I see they appear with two different email addresses (one is invalid of course):

```
Author: Nicolas Gras <nicolasgras01@gmail.com>
Author: Nicolas Gras <nicolas@system76-pc.localdomain>
```

You have to sign the CLA with both to make the checks green, even though the second one is obviously a local one.

- I think @bpintea referred to the content of `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff` (the `enum Part` and some logic) when he asked about common parts that can be extracted to QL.
IMHO it's not mandatory, we could also keep the two things separated and have this PR self-contained in ESQL module, but in that case I'd avoid to move DateTimeField to QL

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nicolasgras committed:
Remove date_diff tests from docs.csv-spec file

nicolasgras committed:
Merge branch 'main' into esql/create-date-diff-function

nicolasgras commented:
Thank you @luigidellaquila for your review.

> You have to sign the CLA with both to make the checks green, even though the second one is obviously a local one.  

I have rebased my commits on top of the branch and updated the author of them, then the CI does not complain anymore about CLA signature.

>  I think @bpintea referred to the content of `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff` (the `enum Part` and some logic) when he asked about common parts that can be extracted to QL.
>       IMHO it's not mandatory, we could also keep the two things separated and have this PR self-contained in ESQL module, but in that case I'd avoid to move DateTimeField to QL.

@bpintea , what do you think about it?
After your review I updated and cleant `org.elasticsearch.xpack.esql.expression.function.scalar.date.DateDiff.Part`, even if it is similar to `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff.Part`, they are a little different now.
And do you think we should move back `DateTimeField` to ESQL?

bpintea commented:
> @bpintea , what do you think about it?

In some cases sharing code between other *ql modules within the `ql` one makes things restrictive/tricky. If the code is very similar, it makes sense, however. But I see you've used some SDK functionality over custom implementation, which is preferred, and if that covers the needs for this PR, I'm OK containing the changes to the `esql` module, as @luigidellaquila suggested. (We can extract these changes in QL and update SQL later if needed.)

nicolasgras committed:
Rollback: DateTimeField interface in esql plugin

nicolasgras commented:
@bpintea and @luigidellaquila, I have moved back `DateTimeField` interface into **ESQL**.
An issue could be created later to extract some code in **QL** and rework some parts in **SQL** if needed.
Is it ok for you?

bpintea reviewed (commented):
I've left two notes, otherwise it LGTM.

I guess with SQL's DateDiff we've follow other implementations returning int and we follow suite also here. This is OK, though we might want to update it, if we'll add support for nano timestamps (which we don't have yet), in which case an int will not cover even 3s.

nicolasgras committed:
Update code after review

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,219 @@
+                    dateTimeUnit
+                );
+                if (similar.isEmpty() == false) {
+                    errorMessage = String.format(
```
ok, I have updated the code

nicolasgras commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:
```diff
@@ -0,0 +1,219 @@
+                Part datePartField = Part.resolve(((BytesRef) unit.fold()).utf8ToString());
+                return new DateDiffConstantEvaluator.Factory(source(), datePartField, startTimestampEvaluator, endTimestampEvaluator);
+            } catch (IllegalArgumentException e) {
+                throw new EsqlIllegalArgumentException(e, "invalid unit format for [{}]: {}", sourceText(), e.getMessage());
```
ok, I have updated the type of Exception thrown here to `InvalidArgumentException`

bpintea reviewed (approved):
Thanks, LGTM.

luigidellaquila reviewed (approved):
Thank you very much @nicolasgras, LGTM

DaveCTurner commented:
Unfortunately this commit broke the build (checkstyle violations) so I reverted it in 943b2eae70e7395ddd6b4a8c4ef36972082ab018. Luigi will open a PR to reinstate it.

luigidellaquila commented:
https://github.com/elastic/elasticsearch/pull/104118
